src/deploy/functions/cel.ts

```diff
@@ -0,0 +1,176 @@
+  if (typeof value !== wantType) {
+    throw new FirebaseError(
+      "CEL identity expression '" + expr + "' resulted in illegal type coercion"
+    );
+  }
+  return value;
+}
+
+function resolveEquality(expr: EqualityExpression, params: Record<string, Literal>): boolean {
+  const match = equalityRegexp.exec(expr);
+  if (!match) {
+    throw new FirebaseError("malformed CEL equality expression '" + expr + "'");
+  }
+
+  const lhs = params[match[1]];
+  let rhs: Literal;
+  if (typeof lhs === "undefined") {
+    throw new FirebaseError(
+      "CEL equality expression LHS '" + match[1] + "' was not resolvable to a param"
+    );
+  } else if (typeof lhs === "string") {
+    rhs = resolveLiteral("string", match[2]);
+  } else if (typeof lhs === "number") {
+    rhs = resolveLiteral("number", match[2]);
+  } else if (typeof lhs === "boolean") {
+    rhs = resolveLiteral("boolean", match[2]);
+  } else {
+    assertExhaustive(lhs);
+  }
+
+  return lhs === rhs;
+}
+
+function resolveTernary(
+  wantType: L,
+  expr: TernaryExpression,
+  params: Record<string, Literal>
+): Literal {
+  const match = ternaryRegexp.exec(expr);
+  if (!match) {
+    throw new FirebaseError("malformed CEL ternary expression '" + expr + "'");
+  }
+
+  // left-hand side of the ternary must be a params.FIELD, supporting any type
+  // right-hand side must be a literal, not of type T but of the same type as the LHS
+  const lhs = params[match[1]];
+  let rhs: Literal;
+  if (typeof lhs === "undefined") {
+    throw new FirebaseError(
+      "CEL ternary expression LHS params.'" + match[1] + "' was not resolvable to a param"
+    );
+  } else if (typeof lhs === "string") {
+    rhs = resolveLiteral("string", match[2]);
+  } else if (typeof lhs === "number") {
+    rhs = resolveLiteral("number", match[2]);
+  } else if (typeof lhs === "boolean") {
+    rhs = resolveLiteral("boolean", match[2]);
+  } else {
+    assertExhaustive(lhs);
+  }
+
+  if (lhs === rhs) {
+    return resolveParamOrLiteral(wantType, match[3], params);
+  } else {
+    return resolveParamOrLiteral(wantType, match[4], params);
+  }
+}
+
+function resolveParamOrLiteral(
+  wantType: L,
+  field: string,
+  params: Record<string, Literal>
+): Literal {
+  const match = paramRegexp.exec(field);
+  if (!match) {
+    return resolveLiteral(wantType, field);
+  }
+  const paramValue = params[match[1]];
+  if (typeof paramValue === "undefined") {
+    throw new FirebaseError("CEL param field '" + field + "' was not provided");
+  }
+  if (typeof paramValue !== wantType) {
+    throw new FirebaseError("CEL param field '" + field + "' resulted in illegal type coercion");
+  }
+  return paramValue;
+}
+
+function resolveLiteral(wantType: L, value: string): Literal {
+  if (paramRegexp.exec(value)) {
+    throw new FirebaseError(
+      "CEL tried to evaluate param." + value + " in a context which only permits literal values"
+    );
+  }
+
+  if (wantType === "number") {
+    if (isNaN(parseInt(value))) {
```

Comment:
don't parseInt. Technically speaking we should probably be figuring out how to disambiguate between ints and floats as params since they're different types. At minimum we shouldn't parseInt because `cpu` can be a fractional amount.